File: codes/kotlin/object_entries.kt
@Suppress("UNUSED_VARIABLE", "UNCHECKED_CAST", "USELESS_CAST")

fun main() {
    fun jsonStringify(anything: Any? = null, pretty: Boolean = false, indent: String = "    "): String {
        var indentLevel = 0
        fun jsonStringifyInner(anythingInner: Any?, indentInner: String): String {
            if (anythingInner == null) return "null"
            if (anythingInner is String) return "\"${anythingInner}\""
            if (anythingInner is Number || anythingInner is Boolean) return "${anythingInner}"
            if (anythingInner is MutableList<*>) {
                if (anythingInner.size == 0) return "[]"
                indentLevel += 1
                var result = (if (pretty == true) "[\n${indentInner.repeat(indentLevel)}" else "[")
                for ((arrayItemIndex, arrayItem) in anythingInner.withIndex()) {
                    result += jsonStringifyInner(arrayItem, indentInner)
                    if ((arrayItemIndex + 1) != anythingInner.size) result += (if (pretty == true) ",\n${indentInner.repeat(indentLevel)}" else ", ")
                }
                indentLevel -= 1
                result += (if (pretty == true) "\n${indentInner.repeat(indentLevel)}]" else "]")
                return result
            }
            if (anythingInner is MutableMap<*, *>) {
                if (anythingInner.entries.size == 0) return "{}"
                indentLevel += 1
                var result = (if (pretty == true) "{\n${indentInner.repeat(indentLevel)}" else "{")
                anythingInner.entries.forEachIndexed { objectEntryIndex, objectEntry ->
                    val objectKey = objectEntry.key
                    val objectValue = objectEntry.value
                    result += "\"${objectKey}\": ${jsonStringifyInner(objectValue, indentInner)}"
                    if ((objectEntryIndex + 1) != anythingInner.entries.size) result += (if (pretty == true) ",\n${indentInner.repeat(indentLevel)}" else ", ")
                }
                indentLevel -= 1
                result += (if (pretty == true) "\n${indentInner.repeat(indentLevel)}}" else "}")
                return result
            }
            return "null"
        }
        return jsonStringifyInner(anything, indent)
    }

    // There's no JavaScript-like Object.entries() in Kotlin.
    // But, we can create our own function to mimic it in Kotlin.

    fun objectEntries(anObject: MutableMap<String, Any?>): MutableList<Any?> {
        // JavaScript-like Object.entries() function
        val newArray = mutableListOf<Any?>()
        anObject.entries.forEachIndexed { _, objectEntry ->
            val objectKey = objectEntry.key
            val objectValue = objectEntry.value
            newArray.add(mutableListOf<Any?>(objectKey, objectValue))
        }
        return newArray
    }

    println("\n// JavaScript-like Object.entries() in Kotlin MutableMap")

    val friend = mutableMapOf<String, Any?>(
        "name" to "Alisa",
        "country" to "Finland",
        "age" to 25
    )
    println("friend: ${jsonStringify(friend, pretty = true)}")

    println("// using JavaScript-like Object.entries() function \"objectEntries\"")

    println("friend entries: ${jsonStringify(objectEntries(friend))}")
    // friend entries: [["name", "Alisa"], ["country", "Finland"], ["age", 25]]
}

File: codes/julia/object_entries.jl
function json_stringify(anything; pretty::Bool = false, indent::String = "    ")::String
    indent_level::Any = 0
    function json_stringify_inner(anything_inner, indent_inner::String)::String
        if (anything_inner === nothing) return "null" end
        if (isa(anything_inner, AbstractString) === true) return "\"$(anything_inner)\"" end
        if ((isa(anything_inner, Number) === true) || (isa(anything_inner, Bool) === true)) return "$(anything_inner)" end
        if (isa(anything_inner, Array) === true)
            if (length(anything_inner) == 0) return "[]" end
            indent_level += 1
            result_array::Any = ((pretty === true) ? "[\n$(repeat(indent_inner, indent_level))" : "[")
            for (array_item_index, array_item) in enumerate(anything_inner)
                result_array *= json_stringify_inner(array_item, indent_inner)
                if (array_item_index !== length(anything_inner)) result_array *= ((pretty === true) ? ",\n$(repeat(indent_inner, indent_level))" : ", ") end
            end
            indent_level -= 1
            result_array *= ((pretty === true) ? "\n$(repeat(indent_inner, indent_level))]" : "]")
            return result_array
        end
        if (isa(anything_inner, Dict) === true)
            if (length(anything_inner) == 0) return "{}" end
            indent_level += 1
            result_object::Any = ((pretty === true) ? "{\n$(repeat(indent_inner, indent_level))" : "{")
            for (object_entry_index, (object_key, object_value)) in enumerate(pairs(anything_inner))
                result_object *= "\"$(object_key)\": $(json_stringify_inner(object_value, indent_inner))"
                if (object_entry_index !== length(anything_inner)) result_object *= ((pretty === true) ? ",\n$(repeat(indent_inner, indent_level))" : ", ") end
            end
            indent_level -= 1
            result_object *= ((pretty === true) ? "\n$(repeat(indent_inner, indent_level))}" : "}")
            return result_object
        end
        return "null"
    end
    return json_stringify_inner(anything, indent)
end

# There's no JavaScript-like Object.entries() in Julia.
# But, we can use Julia Array comprehension and Julia built-in function "pairs" to mimic it in Julia.

function object_entries(an_object)
    # JavaScript-like Object.entries() function
    new_array::Any = Any[]
    for (object_entry_index, (object_key, object_value)) in enumerate(pairs(an_object))
        new_array = Any[new_array..., Any[object_key, object_value]]
    end
    return new_array
end

println("\n# JavaScript-like Object.entries() in Julia Dictionary")

friend::Any = Dict{String, Any}(
    "name" => "Alisa",
    "country" => "Finland",
    "age" => 25
)
println("friend: ", json_stringify(friend, pretty=true))

println("# using Julia Array comprehension and Julia built-in function \"pairs\"")

println("friend entries: ", json_stringify(Array{Any}([[object_key, object_value] for (object_key, object_value) in pairs(friend)])))
# friend entries: [["name", "Alisa"], ["country", "Finland"], ["age", 25]]

println("# using Julia Array comprehension")

println("friend entries: ", json_stringify(Array{Any}([[object_key, object_value] for (object_key, object_value) in friend])))
# friend entries: [["name", "Alisa"], ["country", "Finland"], ["age", 25]]

println("# using JavaScript-like Object.entries() function \"object_entries\"")

println("friend entries: ", json_stringify(Array{Any}(object_entries(friend))))
# friend entries: [["name", "Alisa"], ["country", "Finland"], ["age", 25]]
